workbench-core/datasets/src/dataSetService.ts

```diff
@@ -245,16 +245,32 @@ export class DataSetService {
    *
    * @returns an array of DataSet objects.
    */
-  public async listDataSets(authenticatedUser: { id: string; roles: string[] }): Promise<DataSet[]> {
+  public async listDataSets(authenticatedUser: AuthenticatedUser): Promise<DataSet[]> {
     const metadata: Metadata = {
       actor: authenticatedUser,
       action: this.listDataSets.name,
       source: {
         serviceName: DataSetService.name
       }
     };
+
     try {
-      const response = await this._dbProvider.listDataSets();
+      const response: DataSet[] = [];
+      const allDatasets = await this._dbProvider.listDataSets();
+
+      const allDatasetsPermissions = await Promise.all(
+        allDatasets.map((dataset) =>
+          this._getAuthenticatedUserDatasetPermissions(authenticatedUser, dataset.id)
+        )
+      );
+
+      allDatasetsPermissions.forEach((permission, index) => {
+        if (permission.length) {
+          //eslint-disable-next-line security/detect-object-injection
+          response.push(allDatasets[index]);
```

Comment:
We can only return a dataset here if the permission is for action 'READ' and effect 'ALLOW'. If any permission is a 'DENY'/'READ', it takes precedent over the 'ALLOW' and we do not return it.

This probably needs to be broken out into a common helper that can handle a single dataset so we get it right for external endpoints (which should be updated in a separate ticket) too.

I know we don't directly support having an UPDATE permission without a READ, we don't support CREATE or DELETE permission, and we don't directly support the DENY effect, but because this is authz and can result in potential information disclosure, we need to do it right!

The other option is we throw an exception and return nothing if we hit an unsupported permission, but not sure if that is really the best route here.

Reply:
The functions exposed through the `DataSetsAuthorizationPlugin` dont have the concept of `ALLOW` or `DENY`, just `read-only` and `read-write` so I'm not sure how to implement this without a major rework of the plugin